rclcpp_components/src/component_container_mt.cpp

```diff
@@ -22,8 +22,22 @@ int main(int argc, char * argv[])
 {
   /// Component container with a multi-threaded executor.
   rclcpp::init(argc, argv);
+
+  auto options = rclcpp::NodeOptions{}.start_parameter_event_publisher(false);
   auto exec = std::make_shared<rclcpp::executors::MultiThreadedExecutor>();
-  auto node = std::make_shared<rclcpp_components::ComponentManager>(exec);
+  auto node = std::make_shared<rclcpp_components::ComponentManager>(
+    exec, "ComponentManager", options);
+  if (node->has_parameter("thread_num")) {
+    const auto thread_num = node->get_parameter("thread_num").as_int();
+    if (thread_num < std::thread::hardware_concurrency()) {
+      exec.reset();
+      node.reset();
```

Comment:
you should create the node and the executor correctly once instead of doing this.

e.g.
```cpp
size_t thread_num = 0;  // 0 is the default, which let the MultiThreadedExecutor decide
auto node = make_node();
if (has_param("thread_num")) {
  thread_num = get_param("thread_num");
}
auto executor = std::make_shared<rclcpp::executors::MultiThreadedExecutor>(
  rclcpp::ExecutorOptions{}, thread_num);
```

There's a bit of pseudocode there, but that's the idea.

Reply:
It is necessary to reset node and executor because node has to get an executor as an argument.

Reply:
you can first read the parameter and then create the executor and node correctly; instead of first creating the node/executor, then reading the parameter, and then resetting them if needed.

Reply:
oh, I see that the problem is that you can't read the parameter before creating the node ....
let me see how to fix that

Reply:
I think we can use two step initialization in `ComponentManager`, e.g. add a `set_executor()` method to it.
Then you can create the `ComponentManager` instance (initialized with `executor = nullptr`), read the param, and then call `set_executor()`.

Reply:
@ivanpauno fixed at https://github.com/ros2/rclcpp/pull/1708/commits/fd648c22ef2d4a5294954e7e71d0691ff0eddf8b